### PR TITLE
Change load order of auth backends so that we can throw an exception …

### LINKED ISSUE #25362: Cannot trust servers with Federation app in oC 9.1 RC1 version
### Steps to reproduce

1. Configure two certified oC servers server1 and server2
2. In both servers, modify in /opt/owncloud/config.php the following line
e.g 'overwrite.cli.url' => 'https://server1' and 'https://server2' respectively
3. In admin menu, in Federation section, we write down the url of the other server
4. Back to the CLI, we run the following commands in /opt/owncloud:
```
sudo -u www-data php cron.php (on server1, server2, server1, server2)
sudo -u www-data ./occ dav:sync-system-addressbook (on server1 a server2) 
sudo -u www-data ./occ federation:sync-addressbooks (on server1, server2)
```

### Expected behaviour

After that the address books should be synced and auto complete should work. Trusted server gauge in webUI coloured in green

### Actual behaviour

sudo -u www-data ./occ federation:sync-addressbooks
    1 [->--------------------------]Error while syncing https://docker.oc.solidgear.es:58359 : Unauthorized

Note: It worked fine in oC 9.1.0. pre alpha version
### Server configuration
**Operating system**:
Ubuntu 14.04

**Web server:**
Apache

**Database:**
MySQL

**PHP version:**
5.5.9

**ownCloud version:"9.1.0.8","versionstring":"9.1.0 beta 2","edition":"Enterprise"

**Updated from an older ownCloud or fresh install:**
Fresh

**Are you using external storage, if yes which one:** 
No

**Are you using encryption:** 
No

**Logs**
```
```
### Client configuration

**browser**

Firefox

### PR DESCRIPTION
**Pull request by DeepDiver1975**
…in OCA\DAV\Connector\Sabre\Auth - fixes #25362

Alternative approach to https://github.com/owncloud/core/pull/25474

@davitol please test federation with this fix - THX

### CODE REVIEW

**Comment by davitol** [comment]
:-1: It did not WFM

```
{"reqId":"DRQKxGEA3ErHQk3Zlcpo","remoteAddr":"","app":"federation","message":"Exception: {\"Exception\":\"GuzzleHttp\\\\Exception\\\\ClientException\",\"Message\":\"Client error response [url] http:\\\/\\\/docker.oc.solidgear.es:58662\\\/ocs\\\/v2.php\\\/apps\\\/federation\\\/api\\\/v1\\\/shared-secret?format=json&url=http%3A%2F%2Fdocker.oc.solidgear.es%3A58665&token=NSV6Alq1hktovjD4 [status code] 400 [reason phrase] Bad Request\",\"Code\":400,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Subscriber\\\/HttpError.php(33): GuzzleHttp\\\\Exception\\\\RequestException::create(Object(GuzzleHttp\\\\Message\\\\Request), Object(GuzzleHttp\\\\Message\\\\Response))\\n#1 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Event\\\/Emitter.php(109): GuzzleHttp\\\\Subscriber\\\\HttpError->onComplete(Object(GuzzleHttp\\\\Event\\\\CompleteEvent), 'complete')\\n#2 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(91): GuzzleHttp\\\\Event\\\\Emitter->emit('complete', Object(GuzzleHttp\\\\Event\\\\CompleteEvent))\\n#3 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(132): GuzzleHttp\\\\RequestFsm->__invoke(Object(GuzzleHttp\\\\Transaction))\\n#4 \\\/opt\\\/owncloud\\\/3rdparty\\\/react\\\/promise\\\/src\\\/FulfilledPromise.php(25): GuzzleHttp\\\\RequestFsm->GuzzleHttp\\\\{closure}(Array)\\n#5 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/ringphp\\\/src\\\/Future\\\/CompletedFutureValue.php(55): React\\\\Promise\\\\FulfilledPromise->then(Object(Closure), NULL, NULL)\\n#6 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Message\\\/FutureResponse.php(43): GuzzleHttp\\\\Ring\\\\Future\\\\CompletedFutureValue->then(Object(Closure), NULL, NULL)\\n#7 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(135): GuzzleHttp\\\\Message\\\\FutureResponse::proxy(Object(GuzzleHttp\\\\Ring\\\\Future\\\\CompletedFutureArray), Object(Closure))\\n#8 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(165): GuzzleHttp\\\\RequestFsm->__invoke(Object(GuzzleHttp\\\\Transaction))\\n#9 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(125): GuzzleHttp\\\\Client->send(Object(GuzzleHttp\\\\Message\\\\Request))\\n#10 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/Http\\\/Client\\\/Client.php(136): GuzzleHttp\\\\Client->get('http:\\\/\\\/docker.o...', Array)\\n#11 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(156): OC\\\\Http\\\\Client\\\\Client->get('http:\\\/\\\/docker.o...', Array)\\n#12 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/Job.php(52): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->run(Array)\\n#13 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(134): OC\\\\BackgroundJob\\\\Job->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#14 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(119): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->parentExecute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#15 \\\/opt\\\/owncloud\\\/cron.php(122): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#16 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Exception\\\/RequestException.php\",\"Line\":89}","level":3,"time":"2016-07-14T06:47:32+00:00","method":"--","url":"--","user":"--"}
{"reqId":"BIcBYXLraogo6pzF3u2O","remoteAddr":"","app":"federation","message":"Exception: {\"Exception\":\"GuzzleHttp\\\\Exception\\\\ClientException\",\"Message\":\"Client error response [url] http:\\\/\\\/docker.oc.solidgear.es:58662\\\/ocs\\\/v2.php\\\/apps\\\/federation\\\/api\\\/v1\\\/shared-secret?format=json&url=http%3A%2F%2Fdocker.oc.solidgear.es%3A58665&token=NSV6Alq1hktovjD4 [status code] 400 [reason phrase] Bad Request\",\"Code\":400,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Subscriber\\\/HttpError.php(33): GuzzleHttp\\\\Exception\\\\RequestException::create(Object(GuzzleHttp\\\\Message\\\\Request), Object(GuzzleHttp\\\\Message\\\\Response))\\n#1 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Event\\\/Emitter.php(109): GuzzleHttp\\\\Subscriber\\\\HttpError->onComplete(Object(GuzzleHttp\\\\Event\\\\CompleteEvent), 'complete')\\n#2 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(91): GuzzleHttp\\\\Event\\\\Emitter->emit('complete', Object(GuzzleHttp\\\\Event\\\\CompleteEvent))\\n#3 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(132): GuzzleHttp\\\\RequestFsm->__invoke(Object(GuzzleHttp\\\\Transaction))\\n#4 \\\/opt\\\/owncloud\\\/3rdparty\\\/react\\\/promise\\\/src\\\/FulfilledPromise.php(25): GuzzleHttp\\\\RequestFsm->GuzzleHttp\\\\{closure}(Array)\\n#5 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/ringphp\\\/src\\\/Future\\\/CompletedFutureValue.php(55): React\\\\Promise\\\\FulfilledPromise->then(Object(Closure), NULL, NULL)\\n#6 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Message\\\/FutureResponse.php(43): GuzzleHttp\\\\Ring\\\\Future\\\\CompletedFutureValue->then(Object(Closure), NULL, NULL)\\n#7 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(135): GuzzleHttp\\\\Message\\\\FutureResponse::proxy(Object(GuzzleHttp\\\\Ring\\\\Future\\\\CompletedFutureArray), Object(Closure))\\n#8 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(165): GuzzleHttp\\\\RequestFsm->__invoke(Object(GuzzleHttp\\\\Transaction))\\n#9 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(125): GuzzleHttp\\\\Client->send(Object(GuzzleHttp\\\\Message\\\\Request))\\n#10 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/Http\\\/Client\\\/Client.php(136): GuzzleHttp\\\\Client->get('http:\\\/\\\/docker.o...', Array)\\n#11 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(156): OC\\\\Http\\\\Client\\\\Client->get('http:\\\/\\\/docker.o...', Array)\\n#12 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/Job.php(52): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->run(Array)\\n#13 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(134): OC\\\\BackgroundJob\\\\Job->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#14 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(119): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->parentExecute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#15 \\\/opt\\\/owncloud\\\/cron.php(122): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#16 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Exception\\\/RequestException.php\",\"Line\":89}","level":3,"time":"2016-07-14T06:47:39+00:00","method":"--","url":"--","user":"--"}
{"reqId":"\/B3fEwLN67o82gdMl0mH","remoteAddr":"","app":"federation","message":"Exception: {\"Exception\":\"GuzzleHttp\\\\Exception\\\\ClientException\",\"Message\":\"Client error response [url] http:\\\/\\\/docker.oc.solidgear.es:58662\\\/ocs\\\/v2.php\\\/apps\\\/federation\\\/api\\\/v1\\\/shared-secret?format=json&url=http%3A%2F%2Fdocker.oc.solidgear.es%3A58665&token=NSV6Alq1hktovjD4 [status code] 400 [reason phrase] Bad Request\",\"Code\":400,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Subscriber\\\/HttpError.php(33): GuzzleHttp\\\\Exception\\\\RequestException::create(Object(GuzzleHttp\\\\Message\\\\Request), Object(GuzzleHttp\\\\Message\\\\Response))\\n#1 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Event\\\/Emitter.php(109): GuzzleHttp\\\\Subscriber\\\\HttpError->onComplete(Object(GuzzleHttp\\\\Event\\\\CompleteEvent), 'complete')\\n#2 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(91): GuzzleHttp\\\\Event\\\\Emitter->emit('complete', Object(GuzzleHttp\\\\Event\\\\CompleteEvent))\\n#3 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(132): GuzzleHttp\\\\RequestFsm->__invoke(Object(GuzzleHttp\\\\Transaction))\\n#4 \\\/opt\\\/owncloud\\\/3rdparty\\\/react\\\/promise\\\/src\\\/FulfilledPromise.php(25): GuzzleHttp\\\\RequestFsm->GuzzleHttp\\\\{closure}(Array)\\n#5 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/ringphp\\\/src\\\/Future\\\/CompletedFutureValue.php(55): React\\\\Promise\\\\FulfilledPromise->then(Object(Closure), NULL, NULL)\\n#6 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Message\\\/FutureResponse.php(43): GuzzleHttp\\\\Ring\\\\Future\\\\CompletedFutureValue->then(Object(Closure), NULL, NULL)\\n#7 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/RequestFsm.php(135): GuzzleHttp\\\\Message\\\\FutureResponse::proxy(Object(GuzzleHttp\\\\Ring\\\\Future\\\\CompletedFutureArray), Object(Closure))\\n#8 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(165): GuzzleHttp\\\\RequestFsm->__invoke(Object(GuzzleHttp\\\\Transaction))\\n#9 \\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Client.php(125): GuzzleHttp\\\\Client->send(Object(GuzzleHttp\\\\Message\\\\Request))\\n#10 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/Http\\\/Client\\\/Client.php(136): GuzzleHttp\\\\Client->get('http:\\\/\\\/docker.o...', Array)\\n#11 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(156): OC\\\\Http\\\\Client\\\\Client->get('http:\\\/\\\/docker.o...', Array)\\n#12 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/BackgroundJob\\\/Job.php(52): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->run(Array)\\n#13 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(134): OC\\\\BackgroundJob\\\\Job->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#14 \\\/opt\\\/owncloud\\\/apps\\\/federation\\\/lib\\\/BackgroundJob\\\/GetSharedSecret.php(119): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->parentExecute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#15 \\\/opt\\\/owncloud\\\/cron.php(122): OCA\\\\Federation\\\\BackgroundJob\\\\GetSharedSecret->execute(Object(OC\\\\BackgroundJob\\\\JobList), Object(OC\\\\Log))\\n#16 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/3rdparty\\\/guzzlehttp\\\/guzzle\\\/src\\\/Exception\\\/RequestException.php\",\"Line\":89}","level":3,"time":"2016-07-14T06:47:47+00:00","method":"--","url":"--","user":"--"}
```

**Comment by DeepDiver1975** [comment]
Hmmm .... what is in the log in the other Server?

**Comment by davitol** [comment]
Nothing written in owncloud.log

File: apps/dav/lib/Server.php [hunk 1]
```diff
@@ -66,13 +66,16 @@ public function __construct(IRequest $request, $baseUri) {
 		$this->server->setBaseUri($this->baseUri);
 
 		$this->server->addPlugin(new BlockLegacyClientPlugin(\OC::$server->getConfig()));
-		$authPlugin = new Plugin($authBackend, 'ownCloud');
+		$authPlugin = new Plugin();
```
Comment (PVince81) [inline apps/dav/lib/Server.php]: I think you might also need to set the "priority" attribute or something ? The order is not defined by the order of registration as far as I remember.

**Comment by davitol** [comment]
@DeepDiver1975 Tested again twice and it went fine. I could not reproduce the log error again. Shares went fine, so 👍 IMO

**Comment by PVince81** [comment]
@davitol can you also do a quick regression testing of the different auth methods ? token login, etc

File: apps/dav/lib/Server.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/dav/lib/Server.php]: Afaik Prio is on Plugin registration but here we add auth Backends to one Plugin.

**Comment by PVince81** [comment]
I have retested with two factor and when the login is forbidden without token, this still works.
Code looks good :+1: 

Will prepare backport PR.

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/pull/25478

**Comment by davitol** [comment]
@PVince81 Tested login with token in desktop client too and it went fine. :+1:

**Pull request merged**